module/zcommon/zfs_fletcher.c

```diff
@@ -245,11 +214,258 @@ fletcher_4_incremental_byteswap(const void *buf, uint64_t size,
+			best_run_count = run_count;
+			best_run_index = i;
+		}
+
+		/*
+		 * Due to high overhead of gethrtime(), the performance data
+		 * here is inaccurate and much slower than it could be.
+		 * It's fine for our use though because only relative speed
+		 * is concerned.
+		 */
+		stat->value.ui64 = sizeof (databuf) * run_count *
+		    (NANOSEC / bench_ns) >> 20; /* by MB/s */
+	}
+
+	fletcher_4_fastest = fletcher_4_algos[best_run_index];
+
+	mutex_init(&fletcher_4_impl_seletor_lock, NULL, MUTEX_DEFAULT, NULL);
+	fletcher_4_impl_set("fastest");
+
+	fletcher_4_kstat = kstat_create("zcommon", 0, "fletcher_4_bench",
+	    "misc", KSTAT_TYPE_NAMED, ZFS_ARRAY_SIZE(fletcher_4_algos),
+	    KSTAT_FLAG_VIRTUAL);
+	if (fletcher_4_kstat != NULL) {
+		fletcher_4_kstat->ks_data = fletcher_4_kstat_data;
+		kstat_install(fletcher_4_kstat);
+	}
+}
+
+void
+fletcher_4_fini(void)
+{
+	mutex_destroy(&fletcher_4_impl_seletor_lock);
+	if (fletcher_4_kstat != NULL) {
+		kstat_delete(fletcher_4_kstat);
+		fletcher_4_kstat = NULL;
+	}
+}
+
 #if defined(_KERNEL) && defined(HAVE_SPL)
+
+static int
+fletcher_4_param_get(char *buffer, struct kernel_param *unused)
+{
+	int i, cnt = 0;
+
+	cnt += sprintf(buffer, "Selectors: ");
+	for (i = 0; i < ZFS_ARRAY_SIZE(selector_names); i++) {
+		cnt += sprintf(buffer + cnt,
+		    fletcher_4_impl_seletor == i ? "[%s] " : "%s ",
+		    selector_names[i]);
+	}
+	cnt += sprintf(buffer + cnt, "\n");
+
+	cnt += sprintf(buffer + cnt, "Algorithms: ");
+	for (i = 0; i < ZFS_ARRAY_SIZE(fletcher_4_algos); i++) {
+		cnt += sprintf(buffer + cnt,
+		    fletcher_4_chosen == fletcher_4_algos[i] ? "[%s] " : "%s ",
+		    fletcher_4_algos[i]->name);
+	}
+
+	return (cnt);
+}
+
+static int
+fletcher_4_param_set(const char *val, struct kernel_param *unused)
+{
+	return (fletcher_4_impl_set(val));
+}
+
+module_param_call(fletcher_4_impl_set,
+    fletcher_4_param_set, fletcher_4_param_get, NULL, 0644);
+MODULE_PARM_DESC(fletcher_4_impl_set, "Select fletcher 4 algorithm to use");
+
+EXPORT_SYMBOL(fletcher_4_init);
+EXPORT_SYMBOL(fletcher_4_fini);
 EXPORT_SYMBOL(fletcher_2_native);
 EXPORT_SYMBOL(fletcher_2_byteswap);
 EXPORT_SYMBOL(fletcher_4_native);
 EXPORT_SYMBOL(fletcher_4_byteswap);
 EXPORT_SYMBOL(fletcher_4_incremental_native);
 EXPORT_SYMBOL(fletcher_4_incremental_byteswap);
 #endif
+
+void
+fletcher_4_native(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	const struct fletcher_4_calls *algo = fletcher_4_impl_get();
+
+	algo->init(zcp);
+	algo->compute(buf, size, zcp);
+	if (algo->fini != NULL)
+		algo->fini(zcp);
+}
+
+void
+fletcher_4_byteswap(const void *buf, uint64_t size, zio_cksum_t *zcp)
+{
+	const struct fletcher_4_calls *algo = fletcher_4_impl_get();
+
+	algo->init(zcp);
+	algo->compute_byteswap(buf, size, zcp);
+	if (algo->fini != NULL)
+		algo->fini(zcp);
+}
+
+void
+fletcher_4_incremental_native(const void *buf, uint64_t size,
+    zio_cksum_t *zcp)
+{
+	fletcher_4_generic(buf, size, zcp);
```

Comment:
Why does this directly call `fletcher_4_generic()` instead of `algo->compute()`?  Is the avx2 version of this not incremental?  Same question for `fletcher_4_incremental_byteswap()` below.